framework/src/fwk_status.c

```diff
@@ -0,0 +1,52 @@
+/*
+ * Arm SCP/MCP Software
+ * Copyright (c) 2019, Arm Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Description:
+ *     Framework status code helpers.
+ */
+
+#include <fwk_macros.h>
+#include <fwk_status.h>
+
+static const char * const error_string[] = {
+    [-FWK_E_PARAM]       = "E_PARAM",
+    [-FWK_E_ALIGN]       = "E_ALIGN",
+    [-FWK_E_SIZE]        = "E_SIZE",
+    [-FWK_E_HANDLER]     = "E_HANDLER",
+    [-FWK_E_ACCESS]      = "E_ACCESS",
+    [-FWK_E_RANGE]       = "E_RANGE",
+    [-FWK_E_TIMEOUT]     = "E_TIMEOUT",
+    [-FWK_E_NOMEM]       = "E_NOMEM",
+    [-FWK_E_PWRSTATE]    = "E_PWRSTATE",
+    [-FWK_E_SUPPORT]     = "E_SUPPORT",
+    [-FWK_E_DEVICE]      = "E_DEVICE",
+    [-FWK_E_BUSY]        = "E_BUSY",
+    [-FWK_E_OS]          = "E_OS",
+    [-FWK_E_DATA]        = "E_DATA",
+    [-FWK_E_STATE]       = "E_STATE",
+    [-FWK_E_INIT]        = "E_INIT",
+    [-FWK_E_OVERWRITTEN] = "E_OVERWRITTEN",
+    [-FWK_E_PANIC]       = "E_PANIC",
+};
+
+static const char * const status_string[] = {
+    [FWK_SUCCESS]        = "SUCCESS",
+    [FWK_PENDING]        = "PENDING",
+};
+
+const char *fwk_status_str(int status)
+{
+    unsigned int error_idx = (unsigned int)(-status);
+    static const char unknown[] = "unknown";
+
+    if ((status < 0) && (error_idx < FWK_ARRAY_SIZE(error_string)))
+        return error_string[error_idx];
+
+    if ((status >= 0) && (status < (int)FWK_ARRAY_SIZE(status_string)))
```

Comment:
`else if`

Reply:
may I ask why an `else` is preferred here? I don't against but I feel it adds no value.

Reply:
(despite this is minor stuff) I meant this:
```C
    if ((status < 0) && (error_idx < FWK_ARRAY_SIZE(error_string)))
        return error_string[error_idx];

    if ((status >= 0) && (status < (int)FWK_ARRAY_SIZE(status_string)))
        return status_string[status];

    return unknown;
```
is easier to catch than this:
```C
    if ((status < 0) && (error_idx < FWK_ARRAY_SIZE(error_string)))
        return error_string[error_idx];
    else if ((status >= 0) && (status < (int)FWK_ARRAY_SIZE(status_string)))
        return status_string[status];
    else
        return unknown;
```

Reply:
I'm happy to leave out `else`, but I would prefer we include specifically the `else if` as it just makes it a bit more obvious that the two conditions are tightly connected (`if not a valid error code, check if valid status code`.

Reply:
option1:
```C
    if ((status < 0) && (error_idx < FWK_ARRAY_SIZE(error_string)))
        return error_string[error_idx];
    else if ((status >= 0) && (status < (int)FWK_ARRAY_SIZE(status_string)))
        return status_string[status];

    return unknown;
```

option2 (misra-like):
```C
    if (status < 0) {
        if (error_idx < FWK_ARRAY_SIZE(error_string))
            return error_string[error_idx];
        else
            return unknown;
    } else {
        if (status < (int)FWK_ARRAY_SIZE(status_string)))
            return status_string[status];
        else
            return unknown;
    }
```

Reply:
The first option is preferable - there's no need to do in three conditions what can be done in two, and we aren't bound by MISRA.

Additionally, there's no need for the manual cast as `size_t` has a greater conversion rank than `int`.

So we should use:

```c
    if ((status < 0) && (error_idx < FWK_ARRAY_SIZE(error_string)))
        return error_string[error_idx];
    else if ((status >= 0) && (status < FWK_ARRAY_SIZE(status_string)))
        return status_string[status];

    return unknown;
```

Reply:
ok, thanks.

Regarding the cast, I think it is needed: comparing a signed and an unsigned type is likely to offend the compiler.